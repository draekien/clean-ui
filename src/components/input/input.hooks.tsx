import { useEffect, useState } from 'react';
import { StatusVariant } from '../../types/variants';

export const useVariantColor = (variant?: StatusVariant) => {
  const [variantColor, setVariantColor] = useState('b-300');
  const [variantHoverColor, setVariantHoverColor] = useState('b-400');

  useEffect(() => {
    switch (variant) {
      case 'success':
        setVariantColor('success-400');
        setVariantHoverColor('success-000');
        break;
      case 'warning':
        setVariantColor('warning-400');
        setVariantHoverColor('warning-000');
        break;
      case 'error':
        setVariantColor('error-400');
        setVariantHoverColor('error-000');
        break;
      default:
        setVariantColor('b-300');
        setVariantHoverColor('b-400');
        break;
    }
  }, [variant]);

  return { variantColor, variantHoverColor, setVariantColor };
};
