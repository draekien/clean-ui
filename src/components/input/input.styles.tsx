import { alpha } from '@theme-ui/color';
import { ThemeUIStyleObject } from '@theme-ui/css';
import { Theme } from 'theme-ui';

export const inputWrapperCss = (fullWidth: boolean): ThemeUIStyleObject => ({
  display: 'flex',
  flexFlow: 'column',
  width: fullWidth ? '100%' : 'auto',
});

export const labelContainerCss: ThemeUIStyleObject = {
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'baseline',
};

export const labelCss: ThemeUIStyleObject = {
  variant: 'text.eyebrow',
  color: 'text-muted',
};

export const inputContainerCss: ThemeUIStyleObject = {
  position: 'relative',
  mt: 'xs',
  '> i': {
    position: 'absolute',
    left: 0,
    p: 'xs',
    border: '1px solid transparent',
    color: 'b-400',
  },
};

export const inputCss = (
  icon: boolean,
  color: string,
  hoverColor: string,
  focusColor: string,
  isDarkMode: boolean,
  disabled: boolean
): ThemeUIStyleObject => {
  const css: ThemeUIStyleObject = {
    py: 'xs',
    px: 'sm',
    pl: icon ? 'xxl' : 'sm',
    pr: 'lg',
    width: '100%',
    variant: 'text.body',
    color: 'text',
    backgroundColor: (t: Theme) => `${alpha('muted', isDarkMode ? 0.7 : 0.2)(t)}`,
    borderRadius: 'sm',
    outline: 'none',
    border: '2px solid',
    borderColor: color,
    transition: 'all 300ms',
  };

  if (disabled) {
    css.cursor = 'not-allowed';
    css.borderColor = isDarkMode ? 'b-600' : 'b-200';
    css.color = 'b-400';
  }

  if (!disabled) {
    css[':hover'] = {
      borderColor: hoverColor,
    };
    css[':focus'] = {
      borderColor: focusColor,
    };
  }

  return css;
};

export const helpTextCss = (variantColor: string): ThemeUIStyleObject => ({
  variant: 'text.small',
  color: variantColor,
});
