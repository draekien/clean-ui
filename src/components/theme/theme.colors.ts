export const colors = {
  text: '#292929',
  background: '#F7F7F7',
  primary: '#EF802F',
  secondary: '#DB5192',
  muted: '#D8D8D8',

  'p-000': '#FCE6D5',
  'p-100': '#F9CCAC',
  'p-200': '#F5B382',
  'p-300': '#F29959',
  'p-400': '#EF802F',

  's-000': '#F8DCE9',
  's-100': '#F1B9D3',
  's-200': '#E997BE',
  's-300': '#E274A8',
  's-400': '#DB5192',

  'b-000': '#F7F7F7',
  'b-100': '#EFEFEF',
  'b-200': '#D8D8D8',
  'b-300': '#B1B1B1',
  'b-400': '#898989',
  'b-500': '#626262',
  'b-600': '#3B3B3B',
  'b-700': '#292929',

  'info-000': '#CDE1EB',
  'info-400': '#066899',

  'success-000': '#D5EADC',
  'success-400': '#2E9852',

  'warning-000': '#FBDBD1',
  'warning-400': '#EA4318',

  'error-000': '#F9CCD4',
  'error-400': '#E00027',

  modes: {
    dark: {
      text: '#F7F7F7',
      background: '#292929',
      primary: '#EF802F',
      secondary: '#DB5192',
      muted: '#3B3B3B',
    },
  },
};
