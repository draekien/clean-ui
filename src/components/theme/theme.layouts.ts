export const layouts = {
  radii: {
    xs: '0.125rem',
    sm: '0.25rem',
    md: '0.5rem',
    lg: '0.75rem',
    xl: '1rem',
    max: 999,
  },
  space: {
    xxs: '0.25rem',
    xs: '0.5rem',
    sm: '0.75rem',
    md: '1rem',
    lg: '1.5rem',
    xl: '2rem',
    xxl: '2.5rem',
  },
  shadows: {
    sm: '0px 4px 4px rgba(61, 61, 61, 0.25), 0px 2px 4px -2px rgba(61, 61, 61, 0.25);',
    md: '0px 4px 8px rgba(61, 61, 61, 0.25), 0px 4px 8px -4px rgba(61, 61, 61, 0.25);',
    lg: '0px 4px 8px rgba(61, 61, 61, 0.25), 0px 16px 32px -16px rgba(61, 61, 61, 0.25)',
    text: '0px 2px 4px rgba(61, 61, 61, 0.25);',
    disabled: '0px 2px 4px -2px rgba(4, 40, 52, 0.25);',
  },
  zIndices: {
    lowest: 0,
    lower: 100,
    low: 200,
    middle: 300,
    high: 400,
    higher: 500,
    highest: 600,
    pagination: 1,
    topMenu: 200,
    dropdown: 300,
    drawer: 400,
    modal: 500,
    toast: 600,
    overlay: 600,
  },
};
