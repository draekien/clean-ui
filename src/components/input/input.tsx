/** @jsxImportSource theme-ui */
import { useColorMode } from '@theme-ui/color-modes';
import React, { InputHTMLAttributes } from 'react';
import { StatusVariant } from '../../types/variants';
import { Icon } from '../icon/icon';
import { useVariantColor } from './input.hooks';
import * as styles from './input.styles';

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  /** the id of the input element */
  inputId: string;
  /** label text to display above the input */
  label: string;
  /** a material icon name for displaying an icon inside the input */
  icon?: string;
  /** help text to display under the input */
  helpText?: string;
  /** the input variant to render, 'info' is not supported */
  variant?: StatusVariant;
  /** should the input take up the full width of the parent container? */
  fullWidth?: boolean;
  /** on change event handler */
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => any;
}

export const Input: React.FC<InputProps> = ({
  label,
  inputId,
  icon,
  helpText,
  variant,
  value,
  fullWidth = false,
  onChange,
  ...rest
}) => {
  const [colorMode] = useColorMode();
  const {
    variantColor,
    variantHoverColor,
    variantFocusColor,
    resetVariantColor,
  } = useVariantColor(colorMode === 'dark', variant);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    onChange && onChange(e);
  };

  return (
    <div sx={styles.inputContainerCss(fullWidth)}>
      <label sx={styles.labelCss} htmlFor={inputId}>
        {label}
      </label>
      <div sx={styles.inputWrapperCss}>
        <input
          sx={styles.inputCss(
            !!icon,
            variantColor,
            variantHoverColor,
            variantFocusColor,
            colorMode === 'dark'
          )}
          onFocus={resetVariantColor}
          id={inputId}
          value={value}
          onChange={handleChange}
          {...rest}
        />
        {icon && <Icon name={icon} />}
      </div>
      {helpText && <small sx={styles.helpTextCss(variantHoverColor)}>{helpText}</small>}
    </div>
  );
};
