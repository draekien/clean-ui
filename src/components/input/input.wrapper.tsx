/** @jsxImportSource theme-ui */
import React from 'react';
import { useColorMode } from '@theme-ui/color-modes';
import { InputLabel } from './input.label';
import { StatusVariant } from '../../types/variants';
import { useStatusColors } from '../../hooks/useStatusColors';

import * as styles from './input.styles';

export interface InputWrapperProps extends React.HTMLAttributes<HTMLDivElement> {
  inputId: string;
  label: string;
  helpText?: string;
  variant?: StatusVariant;
  required?: boolean;
  fullWidth?: boolean;
}

export const InputWrapper: React.FC<InputWrapperProps> = ({
  inputId,
  label,
  helpText,
  variant,
  required,
  fullWidth = false,
  children,
  ...rest
}) => {
  const [colorMode] = useColorMode();
  const { color } = useStatusColors(colorMode === 'dark', variant);

  return (
    <div sx={styles.inputWrapperCss(fullWidth)} {...rest}>
      <InputLabel inputId={inputId} label={label} required={required} />
      {children}
      {helpText && <small sx={styles.helpTextCss(color)}>{helpText}</small>}
    </div>
  );
};
