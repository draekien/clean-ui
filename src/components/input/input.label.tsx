/** @jsxImportSource theme-ui */
import React from 'react';
import { Text } from '../text/text';
import { labelContainerCss } from './input.styles';
import { Label } from '../label/label';
import { colors } from '../theme/theme.colors';

export interface InputLabelProps extends React.HTMLAttributes<HTMLDivElement> {
  inputId: string;
  label: string;
  required?: boolean;
  color?: keyof typeof colors;
}

export const InputLabel: React.FC<InputLabelProps> = ({
  inputId,
  label,
  required,
  color = 'text-muted',
}) => {
  return (
    <div sx={labelContainerCss}>
      <Label htmlFor={inputId} color={color}>
        {label}
      </Label>
      {required && (
        <Text as="span" color="warning-400" variant="small">
          required
        </Text>
      )}
    </div>
  );
};
