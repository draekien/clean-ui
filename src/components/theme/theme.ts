import { Theme as ThemeUiTheme } from '@theme-ui/css';
import { colors } from './theme.colors';
import { layouts } from './theme.layouts';
import { styles } from './theme.styles';
import { typography } from './theme.typography';
import { variants } from './theme.variants';

export type Theme = Partial<{}>;

export const CleanUiTheme: ThemeUiTheme = {
  useCustomProperties: false,
  useBorderBox: true,
  useRootStyles: true,
  useLocalStorage: true,
  breakpoints: ['40em', '56em', '64em'],
  colors,
  styles,
  ...variants,
  ...layouts,
  ...typography,
};
