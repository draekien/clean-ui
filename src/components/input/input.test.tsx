import React from 'react';
import { hexToRgb, themedRender } from '../../helpers/testHelpers';
import { StatusVariant } from '../../types/variants';
import { colors } from '../theme/theme.colors';
import { Input, InputWrapper } from './input';

describe('Input Component', () => {
  it('Should match snapshot', () => {
    const { container } = themedRender(<Input inputId="test" />);

    expect(container).toMatchSnapshot();
  });

  const variants: StatusVariant[] = ['info', 'error', 'warning', 'success'];
  variants.forEach((variant) => {
    it(`Should render with correct styles when variant is ${variant}`, () => {
      const { container } = themedRender(<Input inputId="test" variant={variant} />);

      const styles = window.getComputedStyle(container.querySelector('input') as Element);
      expect(styles.borderColor).toBe(colors[`${variant}-300`]);
    });
  });
});

describe('Input Wrapper Component', () => {
  it('Should match snapshot', () => {
    const { container } = themedRender(
      <InputWrapper inputId="test" label="test">
        <Input inputId="test" />
      </InputWrapper>
    );

    expect(container).toMatchSnapshot();
  });

  const variants: StatusVariant[] = ['info', 'error', 'warning', 'success'];
  variants.forEach((variant) => {
    it(`Should render with correct styles when variant is ${variant}`, () => {
      const { container } = themedRender(
        <InputWrapper inputId="test" label="test" helpText="help" variant={variant}>
          <Input inputId="test" />
        </InputWrapper>
      );

      const styles = window.getComputedStyle(container.querySelector('small') as Element);
      expect(styles.color).toBe(hexToRgb(colors[`${variant}-300`]));
    });
  });
});
