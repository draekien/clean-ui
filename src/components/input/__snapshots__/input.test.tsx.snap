// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component Should match snapshot 1`] = `
<div>
  <div
    class="css-1ikh15u-Input"
  >
    <div
      class="css-10vyt3z-Input"
    >
      <label
        class="css-1t4h8g8-Input"
        for="test"
      >
        test
      </label>
    </div>
    <div
      class="css-19x8k4x-Input"
    >
      <input
        class="css-1wao7ws-Input"
        id="test"
        value=""
      />
    </div>
  </div>
</div>
`;
